control_software/src/paper_gpu/file_conversion.py

```diff
@@ -0,0 +1,349 @@
+        # transform coordinates to ECEF
+        lon, lat = latlon_p.transform_point(easting, northing - lat_corr, utm_p)
+        xyz = uvutils.XYZ_from_LatLonAlt(np.radians(lat), np.radians(lon), elevation)
+        antpos_xyz[antnum, :] = xyz
+        # also save antenna names
+        ant_names[antnum] = ant
+
+    return antpos_xyz, ant_names
+
+
+def make_uvh5_file(filename, metadata_file, data_file):
+    """
+    Make a UVH5 file from a metdata + raw binary data file.
+
+    This function creates a valid UVH5 file from the specified metadata and
+    binary data files. It adds the flags and nsample datasets, and applies
+    bitshuffle compression to the data.
+
+    Parameters
+    ----------
+    filename : str
+        The name of the output file to write.
+    metadata_file : str
+        The name of the metadata file written by the correlator.
+    data_file : str
+        The name of the data file written by the correlator.
+
+    Returns
+    -------
+    None
+    """
+    # get cminfo from redis
+    cminfo = redis_cm.read_cminfo_from_redis(return_as="dict")
+
+    # get antennas positions and names from mc
+    antpos_xyz, ant_names = get_antpos_info()
+    nants_telescope = len(ant_names)
+
+    # read in metadata
+    metadata = read_header_data(metadata_file)
+    t0 = metadata[0]
+    mcnt = metadata[1]
+    nfreq = metadata[2]
+    nstokes = metadata[3]
+    corr_ver = metadata[4]
+    tag = metadata[5]
+    ant_0_array = metadata[6]
+    ant_1_array = metadata[7]
+    time_array = metadata[8]
+    integration_time = metadata[9]
+
+    # make sure metadata are the right size
+    nblts = ant_0_array.shape[0]
+    actual_shapes = np.array(
+        [ant_1_array.shape[0], time_array.shape[0], integration_time.shape[0]]
+    )
+    if np.any(actual_shapes != nblts):
+        raise ValueError(
+            f"one or more bad data shapes; expected {nblts}, got {actual_shapes}"
+        )
+
+    # compute other necessary metadata
+    baseline_array = uvutils.antnums_to_baseline(
+        ant_0_array, ant_1_array, nants_telescope
+    )
+    nbls = len(baseline_array)
+    ant_nums = np.asarray([int(name[2:]) for name in ant_names])
+    antenna_diameters = 14.0 * np.ones((len(ant_names),), dtype=np.float64)
+    cofa_lat_rad = cminfo["cofa_lat"] * np.pi / 180.0
+    cofa_lon_rad = cminfo["cofa_lon"] * np.pi / 180.0
+    altitude = cminfo["cofa_alt"]
+    cofa_xyz = uvutils.XYZ_from_LatLonAlt(cofa_lat_rad, cofa_lon_rad, altitude)
+    antpos_xyz -= cofa_xyz
+    # the uvw calculation will have to change when we turn fringe stopping on
+    uvw_array = uvutils.calc_uvw(
+        use_ant_pos=True,
+        antenna_positions=antpos_xyz,
+        antenna_numbers=ant_nums,
+        ant_1_array=ant_0_array,
+        ant_2_array=ant_1_array,
+        telescope_lat=cofa_lat_rad,
+        telescope_lon=cofa_lon_rad,
+        to_enu=True,
+    )
+
+    # build frequency information
+    bandwidth = 250e6
```

Comment:
Someday we want to derive these from redis.

Reply:
Done in 25c0e2f.